src/dict.h

```diff
@@ -185,6 +164,19 @@ void dictRelease(dict *d);
 dictEntry * dictFind(dict *d, const void *key);
 void *dictFetchValue(dict *d, const void *key);
 int dictResize(dict *d);
+void dictSetKey(dict *d, dictEntry* de, void *key);
+void dictSetNext(dictEntry *de, dictEntry *next);
+void *dictMetadata(dictEntry *de);
+void *dictGetKey(const dictEntry *de);
+void *dictGetVal(const dictEntry *de);
+int64_t dictGetSignedIntegerVal(const dictEntry *de);
+uint64_t dictGetUnsignedIntegerVal(const dictEntry *de);
+double dictGetDoubleVal(const dictEntry *de);
+union dictValue *dictGetValPtr(dictEntry *de);
```

Comment:
it looks odd that we call this one `dictGetValPtr` (returns a union), and `dictGetVal` (returns a pointer),maybe it should be the other way around? or maybe we wanna try dropping this interface that returns a union? it's actually used only in one place, which can be replaced with dictGetDoubleVal

Reply:
Yeah, makes sense. I'll look into it.

Reply:
The name is not that illogical since `dictGetValPtr(de)->val == dictGetVal(de)`.

How about `dictGetValUnionPtr`?

This function is used in macros in dict.h so I can't hide it unless these macros are turned into functions within dict.c.

Reply:
come to think of it, isn't exposing that union a compromise on our abstraction of dictEntry?
i.e. we can't safely do nasty things on the LSB of the pointer since people have direct access to it.

in fact, even if we give people access to that variable as a int64_t, we can argue on whatever or not we can mess with that bit. although in practice dictionaries that use it as an integer don't use it as a pointer and vice versa.

bottom line, i think we must get rid of that union.

Reply:
Only dictEntry is made opaque here. The dict value union is still not opaque. We can do nasty things on the dictEntry pointer, but not on the value pointer.

Of course, if everything within dictEntry abstracted, it would be better. That would require turning the macros dictGetValue etc. into functions (which hopefully are inlined by -O3).

Reply:
In the last commit, I made the union value opaque and replaced the macros with functions.

Regarding the pointer access to the double value in z_set, it is documented as a useful optimization, so I've added a special `dictGetDoubleValPtr` to preserve this:

```
                    /* Update the score with the score of the new instance
                     * of the element found in the current sorted set.
                     *
                     * Here we access directly the dictEntry double
                     * value inside the union as it is a big speedup
                     * compared to using the getDouble/setDouble API. */
                    double *existing_score_ptr = dictGetDoubleValPtr(existing);
                    zunionInterAggregate(existing_score_ptr, score, aggregate);
```

It could be that we can replace it with a get + set and rely on inlining here too. I'm leaving this for later...